File: example/wordcount/src/main/java/com/flaredb/example/WordCount.java
package com.flaredb.example;

import java.util.Arrays;

import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.io.TextIO;
import org.apache.beam.sdk.options.PipelineOptionsFactory;
import org.apache.beam.sdk.transforms.Count;
import org.apache.beam.sdk.transforms.DoFn;
import org.apache.beam.sdk.transforms.Filter;
import org.apache.beam.sdk.transforms.FlatMapElements;
import org.apache.beam.sdk.transforms.MapElements;
import org.apache.beam.sdk.transforms.ParDo;
import org.apache.beam.sdk.values.TypeDescriptors;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.flaredb.runner.FlareRunner;

public class WordCount {
    private static final Logger LOG = LoggerFactory.getLogger(WordCount.class);

    public static void main(String[] args) {
        WordCountPipelineOptions options = PipelineOptionsFactory.fromArgs(args)
                .as(WordCountPipelineOptions.class);

        options.setRunner(FlareRunner.class);
        options.setJobEndpoint("127.0.0.1:8099");
        options.setUberJar(
                "/home/ganesh/flare-db/sdf/flare-db/example/wordcount/target/wordcount-1.0-SNAPSHOT.jar");

        Pipeline p = Pipeline.create(options);

        p.apply("ReadLines", TextIO.read().from("/home/ganesh/flare-db/sdf/flare-db/example/wordcount/para.txt"))
                .apply("Split lines into words", FlatMapElements.into(TypeDescriptors.strings())
                        .via(line -> Arrays.asList(line.split(" "))))
                .apply("Remove empty words", Filter.by(word -> !word.isEmpty()))
                .apply("Count occurrences", Count.perElement())
                .apply("Convert counts to strings", MapElements.into(TypeDescriptors.strings())
                        .via(kv -> kv.getKey() + ": " + kv.getValue()))
                .apply("Log results", ParDo.of(new DoFn<String, Void>() {
                    @ProcessElement
                    public void process(ProcessContext ctx) {
                        LOG.info("Element: " + ctx.element());
                    }
                }));

        p.run();
    }
}

File: examples/python/wordcount/wordcount.py
"""WordCount example pipeline for FlareDB.

Python port of the Java reference at
example/wordcount/src/main/java/com/flaredb/example/WordCount.java: the same
five transforms (read, split, filter, count, format) followed by logging
each result, submitted to a FlareDB Job Service via FlareRunner.

Run (from the repository root, with a FlareDB instance already started via
``flare up``):

    pip install -e runner-sdk/python/flarerunner
    python examples/python/wordcount/wordcount.py

See runner-sdk/python/flarerunner/README.md for how to start FlareDB.
"""

import argparse
import logging

import apache_beam as beam
from apache_beam.io import ReadFromText
from apache_beam.options.pipeline_options import PipelineOptions

from flaredb_runner.flare_runner import FlareRunner

LOG = logging.getLogger(__name__)


class LogResult(beam.DoFn):
  """Logs each formatted count, mirroring the Java example's LOG.info ParDo."""

  def process(self, element):
    LOG.info('Element: %s', element)


def run(argv=None):
  """Builds and submits the WordCount pipeline to a FlareDB Job Service."""
  parser = argparse.ArgumentParser()
  parser.add_argument(
      '--input',
      default='test-data/thirukkural.txt',
      help='Path to the input text file to count words from.')
  parser.add_argument(
      '--job_endpoint',
      default='127.0.0.1:8099',
      help='Address of the FlareDB Job Service to submit the pipeline to.')
  known_args, pipeline_args = parser.parse_known_args(argv)

  pipeline_options = PipelineOptions(pipeline_args, job_endpoint=known_args.job_endpoint)

  with beam.Pipeline(runner=FlareRunner(), options=pipeline_options) as p:
    (
        p
        | 'ReadLines' >> ReadFromText(known_args.input)
        | 'Split lines into words' >> beam.FlatMap(lambda line: line.split(' '))
        | 'Remove empty words' >> beam.Filter(lambda word: word != '')
        | 'Count occurrences' >> beam.combiners.Count.PerElement()
        | 'Convert counts to strings' >> beam.MapTuple(lambda word, count: f'{word}: {count}')
        | 'Log results' >> beam.ParDo(LogResult())
    )


if __name__ == '__main__':
  logging.getLogger().setLevel(logging.INFO)
  run()
